### DISCUSSION TITLE
A testbed project for HUB75 panels with WebUI

### DISCUSSION THREAD

**Discussion by vortigont** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#171] [Show and tell]
Hi all,
I find it so frustrating to code/modify/flash any simple patterns/configuration to test this lib, so I decided to spend some time and make a testbed with an easy to use WebUI. Now on since we have a pretty well destructable object and (almost) all options could be specified run-time or on object creation this makes it even more reasonable to have something more complex than a simple examples. So here is what I came up with - [esp32-hub75-testbed](https://github.com/vortigont/esp32-hub75-testbed) project.
It is based on EmbUI framework and let you create and test almost all features of this lib on the fly via simple WebUI interface.
Currently it is just a barely working sketch with tons of bugs and issues but... just need to know if this could be of any use?

Available options:
 - Matrix DMA engine setup with most of the run-time features tunable
 - Virtual panel setup including orientation
 - brightness control
 - solid color fill
 - gradient fill for 3 colors
 - arbitrary straight lines drawing

It could be build with platformio, a proper FS image required.

Known issues:
 - double buffer setup leads to crash due to a lib incompatibility
 - some unknown WiFi instability once DMA starts
 - tons of it actually

An example screenshots
![Screenshot from 2021-08-18 14-16-38](https://user-images.githubusercontent.com/25991192/129889090-515eafe6-4e36-44b8-b1b5-aadf7138c727.png)
![Screenshot from 2021-08-18 14-13-57](https://user-images.githubusercontent.com/25991192/129889094-936d15fa-2074-4b48-b181-eccb5fd03f02.png)

@mrfaptastic @witnessmenow

**Comment by witnessmenow** [comment]
I think this has the possibility of being really useful, especially if it was combined with the web tools stuff I was looking at, so it would really easy to get a user to install it.

I need to figure out the flashing the filesystem part of the web flash thing, although it should not be too hard

Maybe people would disagree, but I think patterns part is not as useful as the dma part, I think a few pre baked options to display would be all you need

**Reply by vortigont** [reply]
Indeed this web tools project looks pretty fine for end users. Although the idea of a browser messing with connected devices drives me crazy  - flashing, destroying, stealing... whatever you (or 'they'?) like? I've already used a tools like this to reflash Bluetooth device :)
Pre baked patters we do have in a simple examples section - it is fine for quick test if everything is OK, but quite annoying if needed to try-and-error some features. Also it was quite interesting for me to play with gradients with this web controls - it helps to understand how this BCM thing works and it's correlation between refresh rate/panel size/etc...
Another idea - it's a great jump start for any projects involving those panels - one firmware for any panel of any size, all the options you can set via WebIU, etc...
